### PR TITLE
Freebsd

### PR DESCRIPTION
**Pull request by rminnich**
This will take a while.
OK, it builds, and we now have freebsd and Plan 9 in CI again.

### CODE REVIEW

**Comment by rminnich** [comment]
There's more to do, but this is getting big and it is probably time for it to go in.

We have let WAY too many linux assumptions creep in, but we're getting close to undoing them.

**Comment by rminnich** [comment]
u-root
.
.
.
11:04:49 INFO Successfully built "/tmp/initramfs.freebsd_amd64.cpio" (size 9296692 bytes -- 8.9 MiB).
FreeBSD freebsd 14.0-RELEASE FreeBSD 14.0-RELEASE #0 releng/14.0-n265380-f9716eee8ab4: Fri Nov 10 05:57:23 UTC 2023     root@releng1.nyi.freebsd.org:/usr/obj/usr/src/amd64.amd64/sys/GENERIC amd64

**Comment by rminnich** [comment]
u-root all now works

File: cmds/core/mknod/mknod_freebsd.go [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+		mode |= unix.S_IFIFO
+		if len(a) != 2 {
+			return fmt.Errorf("device type %v requires no other arguments", devtype)
+		}
+	default:
+		return fmt.Errorf("device type not recognized: %v", devtype)
+	}
+
+	if err := unix.Mknod(path, mode, uint64(dev)); err != nil {
+		return fmt.Errorf("%q: mode %x: %v", path, mode, err)
```
Comment (binjip978) [inline cmds/core/mknod/mknod_freebsd.go:73]: %w for err

File: pkg/libinit/net_freebsd.go [hunk 2]
```diff
@@ -0,0 +1,37 @@
+// Copyright 2014-2019 the u-root Authors. All rights reserved
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+//go:build linux && !tinygo
```
Comment (binjip978) [inline pkg/libinit/net_freebsd.go:5]: wrong build tag? file suffix `is _freebsd`

File: pkg/mount/mount_freebsd.go [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+	}
+	if len(path) == 0 {
+		return errors.New("path cannot be empty")
+	}
+	var flags int
+	if force {
+		flags |= unix.MNT_FORCE
+	}
+	if err := unix.Unmount(path, flags); err != nil {
+		return fmt.Errorf("umount %q flags %x: %v", path, flags, err)
```
Comment (binjip978) [inline pkg/mount/mount_freebsd.go:43]: %w for error

File: pkg/libinit/net_freebsd.go (hunk 2 above)
Comment (rminnich) [reply pkg/libinit/net_freebsd.go:5]: that file was a mistake, it's gone now. Thanks.

File: pkg/mount/mount_freebsd.go (hunk 3 above)
Comment (rminnich) [reply pkg/mount/mount_freebsd.go]: fixed

File: cmds/core/mknod/mknod_freebsd.go (hunk 1 above)
Comment (rminnich) [reply cmds/core/mknod/mknod_freebsd.go:73]: done

File: pkg/libinit/root_freebsd.go [hunk 4]
```diff
@@ -0,0 +1,393 @@
...
+type Mount struct {
+	Source string
+	Target string
+	FSType string
+	Flags  uintptr
+	Opts   string
+}
+
+func (m Mount) Create() error {
+	return os.ErrInvalid
```
Comment (binjip978) [inline pkg/libinit/root_freebsd.go:81]: We have a lot of Mounts in []Namespace slice and all of them will return an error. There is a plan to rewrite Namespaces slice in future?

Comment (rminnich) [reply pkg/libinit/root_freebsd.go:81]: I need to fix mount, but it got complicated, so I wanted to let it go for the next set of PRs. The main goal of this series is to at least get freebsd in the CI testing.

**Comment by rminnich** [comment]
Mount will be fixed next. I need it for cpu :-)

**Review by binjip978 — APPROVED** [review]

**Pull request merged**